src/analysis/ppx_expand.ml

```diff
@@ -0,0 +1,163 @@
+type ppx_expand_result = Query_protocol.ppx_expand_result option
+
+type ppx_kind =
+  | Expr of Parsetree.expression
+  | Sig_item of Parsetree.signature_item
+  | Str_item of Parsetree.structure_item
+
+let attr = ref None
+let expression = ref []
+let signature = ref []
+let structure = ref []
+let kind = ref None
```

Comment:
You're refactoring led to move these reference at the global scope which is something we really don't want to do :-)

I think what I meant was:

1. Have a first iterator which detects the kind of ppx extension we are on: 
```
check_extension : ~parsetree -> pos -> ppx_kind
```
This function can use a reference or a locally defined exception to extract the info from the iterator.

2. And then another function which will have the local references such as `signature` and `structure` but doesn't expose them: 
```
get_ppxed_source : ~ppxed_parsetree pos ppx_kind -> [`Signature of Parsetree.signature | `Structure of Parsetree.structure```

That way the references you're using, and that are implementation details used to extract information from the iterators will be as constrained as possible.


If it's not clear, we can have a call about it !

Reply:
This was a great review. Thanks. 
I've pushed commits with the changes. 

In other to have correct attribute locations for highlighting purposes in VSCode, `check_extension` returns the `ppx_kind * Warnings.loc` and `get_ppxed_source` takes in the `ppx_kind * Warnings.loc`